### DISCUSSION TITLE
ArrayList를 이해하기 전에 Array와 List의 차이를 설명해 봅시다.

### DISCUSSION THREAD

**Discussion by P-C-Space** [NHN-Team-03/ITP#13] [General]
ArrayList는 array와 List를 합친 개념입니다.
그렇다면 ArrayList에서 어떤 주요기능을 빼면 Array가 되고 List가 되는지 설명해 봅시다.

**Comment by damho-lee** [comment]
ArrayList 에서 Random Access를 빼면 List가 되고
ArrayList 에서 자동으로 길이를 늘려주는(가변적으로 길이가 늘어나는) 기능을 빼면 Array가 된다고 생각합니다.

**Comment by P-C-Space** [comment]
Array는 index를 사용, 고정적 크기를 가진다. 따라서 무작위로 접근이 가능하다.|
예를 들어
int [] arrays = new int[100]; 일 때,
랜덤으로 접근이 가능하다.

**Reply by P-C-Space** [reply]
List는 순차적으로 접근해야한다.

**Reply by P-C-Space** [reply]
그러므로 ArrayList는 랜덤 접근이 가능한 List이자 자바에서는 새로운 배열을 생성하여 복사하는 방식이며, 
추가적으로 LinkedList는 순차적으로 접근하는 List이자 자바에서는 Node를 새로 생성하여 이어붙이는 방식이다.

**Comment by uasaha** [comment]
Array는 생성할 때 크기를 할당할 수 있고, 이 크기는 고정적이며 변경 불가하다. (변경하려면 새로운 Array를 선언) 또한 index를 가지고 값을 저장하거나 삭제하는데, 예를 들어 1, 3, 5번 index에 값을 할당하면 0, 2, 4번 index에는 값이 없음에도 불구하고 메모리가 낭비된다. 마지막으로 index는 메모리상에 데이터가 index별로 연속적으로 들어가게 된다.
List는 생성할 때 크키를 할당할 수 있으나 그 크기는 가변적이며 add(), remove()를 통해 내부에 크기가 작아지기도, 커지기도 한다. 또한 add()할 때 메모리의 낭비 없이(비어있는 index가 없음) 차례대로 삽입된다. 마지막으로 Array로 구현한 ArrayList말고, LinkedList는 자식 노드, 부모 노드로 접근하기 때문에 메모리 상에 데이터가 비연속적으로 저장된다.

반박해주세요.

**Reply by uasaha** [reply]
아, LinkedList는 크기를 할당할 수 없습니다.
